### ISSUE TITLE
Contract Upgrade: v63-KREAd

### ISSUE THREAD

**Issue by Chris-Hibbert** [Agoric/Agoric#10392]
## What is the Problem Being Solved?

As mentioned in #8158, we need the ability to restart or replace all vats. This issue is focused on upgrading **v63-KREAd**.

## Description of the Design

When originally written, the KREAd contract was not upgradeable, but #8386 claims that problem has been solved or is straightforward to address.

To close this issue, we need tests that show that
 *  the contract can be upgraded without losing any assets owned by customers, and
 * We've addressed the issue that makes it necessary to call `reviveMarketExitSubscribers` when Zoe is upgraded  #8837
 * an upgrade of KREAd has been added to upgrade.go for the next upgrade cycle.

## Security Considerations

KREAd is not security-critical for our chain.

## Scaling Considerations

Not an issue

## Test Plan

as above

## Upgrade Considerations

See above.

**Comment by Jorge-Lopes** [comment]
## Open Questions:

- Kread contract upgrade
    - Should contract upgrade core-eval execute the `reviveMarketExitSubscribers` as well, or should we keep both processes separated into 2 different scripts?
    - Is there any dependency on the order of execution if we choose to use 2 different core-eval?

**Comment by Jorge-Lopes** [comment]
## Testing Plan

Extend the existing KREAd acceptance tests with the following test cases:

### Test Case: **User assets survive KREAd contract upgrade**

**Steps:**
1. Mint a Character with the default 3 items.
2. Verify that the Character and all 3 items are correctly assigned to the user’s wallet.
3. Upgrade the KREAd contract.
4. Verify that the incarnation number has increased by 1.
5. Ensure that the Character and the 3 items are still properly held by the user post-upgrade.

### Test Case: **Market subscribers survives Zoe contract upgrade**

**Steps:**
1. Sell a Character on the market.
2. Verify that the Character has been removed from the user’s wallet.
3. Ensure the Character is listed on the market.
4. Upgrade the Zoe contract.
5. Verify that the incarnation number has increased by 1.
6. Call `tryExitOffer` with the sellOfferId.
7. Ensure that the sell offer no longer exists in the user’s wallet liveOffers.
8. Verify that the Character is properly held by the user.
9. Ensure that the Character is no longer listed on the market.

**Comment by Jorge-Lopes** [comment]
@Chris-Hibbert could you please provide your opinion on an issue I am facing regarding the core-eval built to execute null upgrade to the KREAd contract.

Thank you in advance.

Discussion link: https://github.com/Agoric/agoric-sdk/discussions/10423

**Comment by Chris-Hibbert** [comment]
> Should contract upgrade core-eval execute the reviveMarketExitSubscribers as well, or should we keep both processes separated into 2 different scripts?

I'd like to find a way to modify the contract so it detects the breakage due to a Zoe restart, and repairs itself automatically. #8714 talks about the problem and potential solutions under `Potential Failure Modes` and `handleExitCharacter failure modes`.  I think adding an error clause to [`getUpdateSince`](https://github.com/agoric-labs/KREAd/blob/54eb7166b8f022cb012f1a01289f1685392e61bb/agoric/contract/src/kreadKit.js#L984) or converting it to `subscribeEach` or `subscribeLatest` would fix it. `isUpgradeDisconnection(error)`  (used in smartWallet and provisionPool) can identify notification failures that are due to Zoe upgrading.

**Comment by Jorge-Lopes** [comment]
# Open questions:

1. Installation Process for Bundles
After making the required changes to the KREAd contract source code and building the respective bundles, what is the recommended process for installing these bundles on-chain before executing the chain software upgrade proposal that will include the KREAd core-eval upgrade?

2. PR to KREAd DCF Fork
Should we create a pull request for the [KREAd DCF fork](https://github.com/DCFoundation/KREAd) containing the updated version of [KREAd](https://github.com/Jorge-Lopes/KREAd/tree/jorge/10392-contract-upgrade)?

3. Package Version and Bundle Size Issue
The issue described in [Discussion #10423](https://github.com/Agoric/agoric-sdk/discussions/10423) was due to the version of the `@agoric/governance` package specified in KREAd, which did not included the changes implemented on this [commit](https://github.com/Agoric/agoric-sdk/pull/9283/commits/84933db9484b5332b1b2a71ca574f0eb536d7905). 
Updating only this package was not feasible due to dependencies between @agoric/governance and other Agoric packages, and so on. 
After updating the necessary packages, the resulting bundle size became quite large, leading to a known issue (see [Discussion #9455](https://github.com/Agoric/agoric-sdk/discussions/9455)).
To mitigate this, I utilized the [resolution field](https://github.com/Jorge-Lopes/KREAd/blob/jorge/10392-contract-upgrade/agoric/package.json#L58-L84) to address both dependency and size issues. 
Is this approach suitable, and is there a preferred version of Agoric packages I should specify?

**Comment by Jorge-Lopes** [comment]
> I'd like to find a way to modify the contract so it detects the breakage due to a Zoe restart, and repairs itself automatically. #8714 talks about the problem and potential solutions under `Potential Failure Modes` and `handleExitCharacter failure modes`. I think adding an error clause to [`getUpdateSince`](https://github.com/agoric-labs/KREAd/blob/54eb7166b8f022cb012f1a01289f1685392e61bb/agoric/contract/src/kreadKit.js#L984) or converting it to `subscribeEach`or`subscribeLatest`would fix it.`isUpgradeDisconnection(error)` (used in smartWallet and provisionPool) can identify notification failures that are due to Zoe upgrading.

The suggested approach is implemented in the following [commit](https://github.com/Jorge-Lopes/KREAd/commit/ba7369500c881e8661401bf2714c41a626031ee0).
I opted to use the `subscribeLatest` method to replace `getUpdateSince` due to the similarity in the lossy consumption of a `subscriber`

**Comment by Jorge-Lopes** [comment]
@Chris-Hibbert to remove the necessity to manually install the new KREAd contract and manifest bundles, I updated the acceptance KREAd test to execute that process during the test execution at the commit [233074e](https://github.com/Agoric/agoric-sdk/commit/233074ec1af0ee95877515eb2630cbd7a2988379).

The bundles that were added to `a3p-integration/proposals/z:acceptance/kreadBundles/` were generated, after the refactoring the KREAd contract, using the available [target](https://github.com/Jorge-Lopes/KREAd/blob/jorge/10392-contract-upgrade/agoric/Makefile#L70-L77) which executes the `agoric run` command with the original core-eval [script](https://github.com/Jorge-Lopes/KREAd/blob/jorge/10392-contract-upgrade/agoric/contract/src/proposal/start-kread-script.js)

Note: this commit is expected to be reverted when a more suitable solution is found

**Comment by Chris-Hibbert** [comment]
> To mitigate this, I utilized the [resolution field](https://github.com/Jorge-Lopes/KREAd/blob/jorge/10392-contract-upgrade/agoric/package.json#L58-L84) to address both dependency and size issues.
Is this approach suitable, and is there a preferred version of Agoric packages I should specify?

That seems reasonable to me.
I've heard that `dev` is the version to use in `a3p-integration`.  (@turadg, please verify this.)

**Comment by turadg** [comment]
> That seems reasonable to me. I've heard that `dev` is the version to use in `a3p-integration`. (@turadg, please verify this.)

Yes, that's what should go in `dependencies`. However it won't matter in `z:accptance` because of the new `yarn link` resolutions: https://github.com/Agoric/agoric-sdk/pull/10396/files#diff-62c4b0398df47ad214b7ba03a0fee4dda43936268f3944c8d916e51c594334f6

**Comment by Jorge-Lopes** [comment]
During the acceptance test execution, after upgrading the KREAd contract, the market is generally functioning as expected. Specifically, for two sell offers placed before the zoe-vat upgrade, the following outcomes were observed:

- For the sell offer that was exited after the upgrade, the associated character was no longer available on the character market.
- For the second sell offer, the character remained available on the market, and a successful buy offer was made for it.

However, an error is being logged immediately after zoe-vat upgrade for each pending sell offer, similar to the chain logs shown below.

@Chris-Hibbert Could this indicate that replacing `getUpdateSince` with `subscribeLatest` in this [commit](https://github.com/Jorge-Lopes/KREAd/commit/ba7369500c881e8661401bf2714c41a626031ee0) was not sufficient to resolve the issue described in Issue #8714 ?

```sh
2024-11-12T12:04:34.635Z SwingSet: vat: v1: evaluateBundleCap { manifestBundleRef: { bundleID: 'b1-09b92c35ba08e8d3625eb04796ab70cd02cf9ef3bfd79e8f256b419afb644e4a9e9e4b06c068abe2ecc2769a53a55024a27851ec7bfeb13e79b3781d0b961597' }, manifestGetterName: 'getManifestForUpgradingZoe', vatAdminSvc: Promise [Promise] {} }
2024-11-12T12:04:34.709Z SwingSet: vat: v1: execute { manifestGetterName: 'getManifestForUpgradingZoe', bundleExports: [ 'getManifestForUpgradingZoe', 'upgradeZoe' ] }
2024-11-12T12:04:34.710Z SwingSet: vat: v1: coreProposal: upgradeZoe
2024-11-12T12:04:34.716Z SwingSet: vat: v1: ZOE BUNDLE ID:  b1-a6e125677dde1b87dbe7e533d30381d5ae7fc49d2df323923ac0e9912fd01149ee885b9f8be7d76451325ca9c6de8518def0eae366118788b68313e61aa7a1f0
2024-11-12T12:04:34.720Z SwingSet: kernel: attempting to upgrade vat v9 from incarnation 3 to source b1-a6e125677dde1b87dbe7e533d30381d5ae7fc49d2df323923ac0e9912fd01149ee885b9f8be7d76451325ca9c6de8518def0eae366118788b68313e61aa7a1f0
2024-11-12T12:04:36.262Z SwingSet: kernel: vat v9 upgraded from incarnation 3 to 4 with source b1-a6e125677dde1b87dbe7e533d30381d5ae7fc49d2df323923ac0e9912fd01149ee885b9f8be7d76451325ca9c6de8518def0eae366118788b68313e61aa7a1f0
2024-11-12T12:04:37.425Z SwingSet: vat: v43: wallet agoric1rwwley550k9mmk6uq6mm6z4udrg8kyuyvfszjk IMMEDIATE OFFER ERROR: { incarnationNumber: 3, name: 'vatUpgraded', upgradeMessage: 'vat upgraded' }
2024-11-12T12:04:37.429Z SwingSet: xsnap: v43: UnhandledPromiseRejectionWarning: (Error#1)
2024-11-12T12:04:37.429Z SwingSet: xsnap: v43: Error#1: Pending undefined could not retry; { incarnationNumber: 3, name: 'vatUpgraded', upgradeMessage: 'vat upgraded' }
2024-11-12T12:04:37.429Z SwingSet: xsnap: v43: Error: Pending "[undefined]" could not retry; (an object)
 at apply ()
 at Error (/bundled-source/.../node_modules/ses/src/error/tame-error-constructor.js:60)
 at makeError (/bundled-source/.../node_modules/ses/src/error/assert.js:352)
 at fail (/bundled-source/.../node_modules/ses/src/error/assert.js:482)
 at onRejected (.../vow/src/watch-utils.js:265)
 at apply ()
 at In "onRejected" method of (WatchUtils retryRejectionPromiseWatcher) (/bundled-source/.../node_modules/@endo/exo/src/exo-tools.js:171)
 at (/bundled-source/.../packages/swingset-liveslots/src/watchedPromises.js:118)
 at ()
```

**Comment by Jorge-Lopes** [comment]
@Chris-Hibbert when you have the availability, could you please provide a feedback regarding the following question:

> 1. Installation Process for Bundles
After making the required changes to the KREAd contract source code and building the respective bundles, what is the recommended process for installing these bundles on-chain before executing the chain software upgrade proposal that will include the KREAd core-eval upgrade?

**Comment by Chris-Hibbert** [comment]
The advice I got was that this upgrade should be tested in KREAd's repo rather than `agoric-sdk/a3p-integration`.

The way to proceed would be to
 * create a PR against `https://github.com/Kryha/KREAd/`, proposing these changes. (I'll find out what we'll need to do to get it reviewed and approved.)
 * Include a new `a3p-integration` subdirectory, copied from `agoric-sdk`.  
 * The proposal would be a core-eval, (s:stake-bid is an example). 
   * The assets (bundles, script, proposal) can be placed directly in a `submission/` directory
 * When it's time to create a core-eval proposal to run on chain (whether devnet, mainnet or other testNet) we'd paste the assets into the [`cosgov`](https://cosgov.org/agoric?network=ollinet) UI. (ask me for details)
 * Once the changes have been placed on-chain, we'd copy the a3p test to our `agoric-3-proposals` repo.

So this follows the process that would be used for any 3rd party code.

**Comment by Chris-Hibbert** [comment]
> During the acceptance test execution

This would be easier to review with a PR and CI test output. Would you provide a link?

**Comment by Jorge-Lopes** [comment]
> This would be easier to review with a PR and CI test output. Would you provide a link?

I opened a draft PR, where the existing acceptance kread test was extended with the both test cases described [here](https://github.com/Agoric/agoric-sdk/issues/10392#issuecomment-2462747183).
The bundles are being installed during test execution, along with the kread contract upgrade and zoe-vat upgrade.
During the execution of the last KREAd test case, 'market survives zoe upgrade' , the logs posted bellow will be printed on console.

PR: https://github.com/Agoric/agoric-sdk/pull/10465

Is this test environment appropriate for reproducing the issue, or would you prefer an alternative setup?

**Comment by Chris-Hibbert** [comment]
> Is this test environment appropriate for reproducing the issue, or would you prefer an alternative setup?

Yep, but it needs `force-integration` before it'll run the docker tests. I added the label.

**Comment by Chris-Hibbert** [comment]
I can't tell whether the error messages here
```
2024-11-13T21:37:37.173Z SwingSet: vat: v43: wallet agoric1rwwley550k9mmk6uq6mm6z4udrg8kyuyvfszjk IMMEDIATE OFFER ERROR: { incarnationNumber: 3, name: 'vatUpgraded', upgradeMessage: 'vat upgraded' }
2024-11-13T21:37:37.188Z SwingSet: xsnap: v43: Error#1: Pending undefined could not retry; { incarnationNumber: 3, name: 'vatUpgraded', upgradeMessage: 'vat upgraded' }
```
are being generated from the new `subscribeLatest` code or the previous code that that's replacing. Would you link to the kread repairs that you're using to build the bundles?

In the past, in similar situations, we've seen the pre-upgrade promises generate error messages that reflect the kernel handling the dangling broken promises. Since the old code is gone, the kernel can't tell what to do with the results from the old promise.  If `watch-utils.js` is only invoked from the new code then this is not in the old code.

Can you tell whether the KREAd code continues to work after this exception? If it's dead following that, then we have a problem, and if it works afterwards, then this might be a spurious error.

**Comment by anilhelvaci** [comment]
Not included in the upgrade process as of today.

**Comment by turadg** [comment]
I hear @otoole-brendan decided this isn't worthwhile. I'm closing and asking @aj-agoric to make sure there's a ticket to clean it up.
